### PR TITLE
depends: Purge libtool archives

### PR DESCRIPTION
**Pull request by dongcarl**
```
We use pkg-config where we can, which generally replaces libtool at a
higher level and does not have the same downsides as libtool. These
archives sit in our depends tree with no purpose and pollute the final
bitcoin build with massive overlinking.
```

See [here](https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Handling_Libtool_Archives) for an explanation of the various problems libtool archives can cause.

Unrelated in every way except in spirit: `-D__LIBTOOL_IS_A_FOOL__`!!

-----

This PR is based on #16041, and therefore should be merged after #16041.

### CODE REVIEW

**Comment by laanwj** [comment]
I had to do this once for an embedded project, because the `la` files were confusing things (as they were generated in a different environment). But I'm not sure here, do you have evidence of "massive overlinking", what is the result of that? Does the build become slower, does the executable become larger? If so, can you show results please?

**Comment by dongcarl** [comment]
@laanwj 

The result is usually `-lstdc++` and various dynamic linking flags we don't want ending up in the `.la` files, and so when we do the final bitcoin build, we're getting extra symbols that aren't necessary in our ELF table. This makes `make -C src check-symbols` fail.

You can expand "Old content no. 1" [here](https://github.com/bitcoin/bitcoin/pull/15277#issue-248219051) for an example.

**Comment by theuni** [comment]
This comes at my request, so concept ACK :)

I suspect it may flesh out a few actual bugs, where we had assumed pkg-config was doing its job.

**Comment by theuni** [comment]
@dongcarl Could you try this https://github.com/theuni/bitcoin/commit/f643224b4ca147f8c127804d1c03237eee345d03 as a test fix for the new failure?

**Comment by dongcarl** [comment]
@theuni Hmmm I'm not encountering failures, but the ones I've encountered before were related to `_XEatDataWords` which is fixed by https://github.com/freedesktop/libXext/commit/bb24f2970f2e425f4df90c9b73d078ad15a73fbb. On my local builds I've just upgraded the package to the earliest version that contained that fix: https://github.com/freedesktop/libXext/releases/tag/libXext-1.3.3

**Comment by theuni** [comment]
@dongcarl Hah, yes, thanks for noticing. I just assumed ours wasn't ancient. Please include the bump with this PR, since it breaks the build without it.

**Comment by dongcarl** [comment]
Looking at the [Travis build logs](https://travis-ci.org/bitcoin/bitcoin/builds/522310539) for https://github.com/bitcoin/bitcoin/pull/15844/commits/27ce43dcd512d1158c7400ac03afbdd1af2d7031, it would seem that cross builds are running into another problem that I assumed to be specific to my particular configuration. Namely, that we need to give the config flag `--enable-malloc0returnsnull`.

I've included that config flag in https://github.com/bitcoin/bitcoin/pull/15844/commits/e5444bbd13a30448f3d811d4f1faef058f18e65a, this seems to be a common problem for cross builds of X things: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=861073

**Comment by fanquake** [comment]
Concept ACK

Looks like there is still at least one issue with the `32-bit + dash [GOAL: install] [GUI: no BIP70]` build [here](https://travis-ci.org/bitcoin/bitcoin/jobs/522320123):
```
checking for QTACCESSIBILITY... yes
checking for QTFB... yes
checking for X11XCB... yes
checking for QTXCBQPA... yes
checking for static Qt plugins: -lqminimal... no
configure: error: Could not resolve: -lqminimal
```

This is the diff of `/lib` for a `master` depends build, vs this PR on macOS, will test on others:

`diffoscope x86_64-apple-darwin18.5.0-master/lib x86_64-apple-darwin18.5.0/lib`:

```diff
@@ -28,37 +28,28 @@
libboost_unit_test_framework-mt.a
libcrypto.a
libdb-4.8.a
libdb.a
libdb_cxx-4.8.a
libdb_cxx.a
libevent.a
-libevent.la
libevent_core.a
-libevent_core.la
libevent_extra.a
-libevent_extra.la
libevent_pthreads.a
-libevent_pthreads.la
libminiupnpc.a
libprotobuf-lite.a
-libprotobuf-lite.la
libprotobuf.a
-libprotobuf.la
-libprotoc.la
libqrencode.a
-libqrencode.la
libqtharfbuzz.a
libqtlibpng.a
libqtpcre2.a
librapidcheck.a
libssl.a
libz.a
libzmq.a
-libzmq.la
pkgconfig
pkgconfig/Qt5AccessibilitySupport.pc
pkgconfig/Qt5CglSupport.pc
pkgconfig/Qt5ClipboardSupport.pc
pkgconfig/Qt5Core.pc
pkgconfig/Qt5DBus.pc
pkgconfig/Qt5DeviceDiscoverySupport.pc
```

**Comment by fanquake** [comment]
After depends building inside a `ubuntu:1804` Docker container.

`diffoscope x86_64-pc-linux-gnu-master/lib x86_64-pc-linux-gnu/lib`:

```diff
@@ -15,25 +15,21 @@
libQt5Network.a
libQt5PrintSupport.a
libQt5ServiceSupport.a
libQt5Test.a
libQt5ThemeSupport.a
libQt5Widgets.a
libQt5XcbQpa.a
-libX11-xcb.la
libX11-xcb.so
libX11-xcb.so.1
libX11-xcb.so.1.0.0
-libX11.la
libX11.so
libX11.so.6
libX11.so.6.3.0
libXau.a
-libXau.la
-libXext.la
libXext.so
libXext.so.6
libXext.so.6.4.0
libboost_chrono-mt.a
libboost_filesystem-mt.a
libboost_prg_exec_monitor-mt.a
libboost_system-mt.a
@@ -42,154 +38,117 @@
libboost_timer-mt.a
libboost_unit_test_framework-mt.a
libcrypto.a
libdb-4.8.a
libdb.a
libdb_cxx-4.8.a
libdb_cxx.a
-libdbus-1.la
libdbus-1.so
libdbus-1.so.3
libdbus-1.so.3.14.11
libevent.a
-libevent.la
libevent_core.a
-libevent_core.la
libevent_extra.a
-libevent_extra.la
libevent_pthreads.a
-libevent_pthreads.la
-libexpat.la
libexpat.so
libexpat.so.1
libexpat.so.1.6.8
-libfontconfig.la
libfontconfig.so
libfontconfig.so.1
libfontconfig.so.1.9.2
-libfreetype.la
libfreetype.so
libfreetype.so.6
libfreetype.so.6.13.0
libminiupnpc.a
libprotobuf-lite.a
-libprotobuf-lite.la
libprotobuf.a
-libprotobuf.la
-libprotoc.la
libqrencode.a
-libqrencode.la
libqtharfbuzz.a
libqtlibpng.a
libqtpcre2.a
librapidcheck.a
libssl.a
-libxcb-composite.la
libxcb-composite.so
libxcb-composite.so.0
libxcb-composite.so.0.0.0
-libxcb-damage.la
libxcb-damage.so
libxcb-damage.so.0
libxcb-damage.so.0.0.0
-libxcb-dpms.la
libxcb-dpms.so
libxcb-dpms.so.0
libxcb-dpms.so.0.0.0
-libxcb-dri2.la
libxcb-dri2.so
libxcb-dri2.so.0
libxcb-dri2.so.0.0.0
-libxcb-dri3.la
libxcb-dri3.so
libxcb-dri3.so.0
libxcb-dri3.so.0.0.0
-libxcb-glx.la
libxcb-glx.so
libxcb-glx.so.0
libxcb-glx.so.0.0.0
-libxcb-present.la
libxcb-present.so
libxcb-present.so.0
libxcb-present.so.0.0.0
-libxcb-randr.la
libxcb-randr.so
libxcb-randr.so.0
libxcb-randr.so.0.1.0
-libxcb-record.la
libxcb-record.so
libxcb-record.so.0
libxcb-record.so.0.0.0
-libxcb-render.la
libxcb-render.so
libxcb-render.so.0
libxcb-render.so.0.0.0
-libxcb-res.la
libxcb-res.so
libxcb-res.so.0
libxcb-res.so.0.0.0
-libxcb-screensaver.la
libxcb-screensaver.so
libxcb-screensaver.so.0
libxcb-screensaver.so.0.0.0
-libxcb-shape.la
libxcb-shape.so
libxcb-shape.so.0
libxcb-shape.so.0.0.0
-libxcb-shm.la
libxcb-shm.so
libxcb-shm.so.0
libxcb-shm.so.0.0.0
libxcb-static.a
-libxcb-sync.la
libxcb-sync.so
libxcb-sync.so.1
libxcb-sync.so.1.0.0
-libxcb-xevie.la
libxcb-xevie.so
libxcb-xevie.so.0
libxcb-xevie.so.0.0.0
-libxcb-xf86dri.la
libxcb-xf86dri.so
libxcb-xf86dri.so.0
libxcb-xf86dri.so.0.0.0
-libxcb-xfixes.la
libxcb-xfixes.so
libxcb-xfixes.so.0
libxcb-xfixes.so.0.0.0
-libxcb-xinerama.la
libxcb-xinerama.so
libxcb-xinerama.so.0
libxcb-xinerama.so.0.0.0
-libxcb-xkb.la
libxcb-xkb.so
libxcb-xkb.so.1
libxcb-xkb.so.1.0.0
-libxcb-xprint.la
libxcb-xprint.so
libxcb-xprint.so.0
libxcb-xprint.so.0.0.0
-libxcb-xtest.la
libxcb-xtest.so
libxcb-xtest.so.0
libxcb-xtest.so.0.0.0
-libxcb-xv.la
libxcb-xv.so
libxcb-xv.so.0
libxcb-xv.so.0.0.0
-libxcb-xvmc.la
libxcb-xvmc.so
libxcb-xvmc.so.0
libxcb-xvmc.so.0.0.0
-libxcb.la
libxcb.so
libxcb.so.1
libxcb.so.1.1.0
libz.a
libzmq.a
-libzmq.la
pkgconfig
pkgconfig/Qt5AccessibilitySupport.pc
pkgconfig/Qt5Core.pc
pkgconfig/Qt5DBus.pc
pkgconfig/Qt5DeviceDiscoverySupport.pc
pkgconfig/Qt5EventDispatcherSupport.pc
pkgconfig/Qt5FbSupport.pc
```

**Comment by dongcarl** [comment]
Ready for re-review. Checks are passing.

**Comment by theuni** [comment]
Thanks, having a look at this.

File: depends/packages/libXext.mk
Comment (theuni) [inline depends/packages/libXext.mk]: I looked into this a few weeks back, and I believe this is auto-detected, but we manage to get it wrong.

Could you please explain exactly why this is necessary? I'm afraid we're accidentally building against system files.

File: depends/packages/expat.mk
Comment (theuni) [inline depends/packages/expat.mk]: I think this is a remnant of the qt4 days. I can't recall why it would've been necessary, but I agree that this is correct now.

File: depends/packages/libXext.mk
Comment (dongcarl) [reply depends/packages/libXext.mk]: I mentioned this a little here: https://github.com/bitcoin/bitcoin/pull/15844#issuecomment-485025030

Also, it's recommended when cross-compiling here: https://www.x.org/wiki/CrossCompilingXorg/

Comment (theuni) [reply depends/packages/libXext.mk]: You're completely right, thank you.

Could you add that link (or c/p the quote) as a comment? It's very helpful to know that it's for bypassing an AC_RUN_IFELSE.

Comment (dongcarl) [reply depends/packages/libXext.mk]: Actually, after looking at libXext's source I believe the stricter `--disable-malloc0returnsnull` works as well (I also confirmed thru a gitian build). I'll update this.

**Comment by dongcarl** [comment]
Updated with some commentary to help future cross-compilation target architectures/libcs.

**Comment by dongcarl** [comment]
Rebased.

File: depends/packages/libXext.mk
Comment (fanquake) [inline depends/packages/libXext.mk]: I've only looked into this briefly, but why does `--disable-malloc0returnsnull` only get passed to `libXext`?

With this change, building depends on Debian I'm seeing:
```
Configuring libX11...
checking whether malloc(0) returns NULL... yes

Configuring libXext...
checking whether malloc(0) returns NULL... no
``` 

Does this matter?

Comment (dongcarl) [reply depends/packages/libXext.mk]: @fanquake Ah, that's the only place where I found problems. Looking at the `libX11` `aclocal.m4`, they haven't updated their `xorg-macros` yet so no problems yet. We should add `--disable-malloc0returnsnull` there as well just to future-proof. You didn't see any other `malloc(0)` checks other than these two?

Comment (fanquake) [reply depends/packages/libXext.mk]: I haven't seen any other `malloc(0)` checks 👍.

**Comment by fanquake** [comment]
tACK 716afd6

`.la`s all seem to be removed, and we pick up a `libexpat.a` because of 716afd6.

Building on macOS 10.14.x:
<details><summary><code>diffoscope x86_64-apple-darwin18.6.0/lib x86_64-apple-darwin18.6.0-716afd6a3/lib</code></summary>

```diff
libboost_unit_test_framework-mt.a
libcrypto.a
libdb-4.8.a
libdb.a
libdb_cxx-4.8.a
libdb_cxx.a
libevent.a
-libevent.la
libevent_core.a
-libevent_core.la
libevent_extra.a
-libevent_extra.la
libevent_pthreads.a
-libevent_pthreads.la
libminiupnpc.a
libprotobuf-lite.a
-libprotobuf-lite.la
libprotobuf.a
-libprotobuf.la
-libprotoc.la
libqrencode.a
-libqrencode.la
libqtharfbuzz.a
libqtlibpng.a
libqtpcre2.a
librapidcheck.a
libssl.a
libz.a
libzmq.a
-libzmq.la
pkgconfig
pkgconfig/Qt5AccessibilitySupport.pc
pkgconfig/Qt5CglSupport.pc
pkgconfig/Qt5ClipboardSupport.pc
pkgconfig/Qt5Core.pc
pkgconfig/Qt5DBus.pc
pkgconfig/Qt5DeviceDiscoverySupport.pc
```
</details>

Building on Debian 9.9:
<details><summary><code>diffscope x86_64-pc-linux-gnu/lib x86_64-pc-linux-gnu-716afd6a3/lib</code></summary>

```diff
libQt5Network.a
libQt5PrintSupport.a
libQt5ServiceSupport.a
libQt5Test.a
libQt5ThemeSupport.a
libQt5Widgets.a
libQt5XcbQpa.a
-libX11-xcb.la
libX11-xcb.so
libX11-xcb.so.1
libX11-xcb.so.1.0.0
-libX11.la
libX11.so
libX11.so.6
libX11.so.6.3.0
libXau.a
-libXau.la
-libXext.la
libXext.so
libXext.so.6
libXext.so.6.4.0
libboost_chrono-mt.a
libboost_filesystem-mt.a
libboost_prg_exec_monitor-mt.a
libboost_system-mt.a
@@ -42,155 +38,115 @@
libboost_timer-mt.a
libboost_unit_test_framework-mt.a
libcrypto.a
libdb-4.8.a
libdb.a
libdb_cxx-4.8.a
libdb_cxx.a
-libdbus-1.la
libdbus-1.so
libdbus-1.so.3
libdbus-1.so.3.14.11
libevent.a
-libevent.la
libevent_core.a
-libevent_core.la
libevent_extra.a
-libevent_extra.la
libevent_pthreads.a
-libevent_pthreads.la
-libexpat.la
-libexpat.so
-libexpat.so.1
-libexpat.so.1.6.8
-libfontconfig.la
+libexpat.a
libfontconfig.so
libfontconfig.so.1
libfontconfig.so.1.9.2
-libfreetype.la
libfreetype.so
libfreetype.so.6
libfreetype.so.6.13.0
libminiupnpc.a
libprotobuf-lite.a
-libprotobuf-lite.la
libprotobuf.a
-libprotobuf.la
-libprotoc.la
libqrencode.a
-libqrencode.la
libqtharfbuzz.a
libqtlibpng.a
libqtpcre2.a
librapidcheck.a
libssl.a
-libxcb-composite.la
libxcb-composite.so
libxcb-composite.so.0
libxcb-composite.so.0.0.0
-libxcb-damage.la
libxcb-damage.so
libxcb-damage.so.0
libxcb-damage.so.0.0.0
-libxcb-dpms.la
libxcb-dpms.so
libxcb-dpms.so.0
libxcb-dpms.so.0.0.0
-libxcb-dri2.la
libxcb-dri2.so
libxcb-dri2.so.0
libxcb-dri2.so.0.0.0
-libxcb-dri3.la
libxcb-dri3.so
libxcb-dri3.so.0
libxcb-dri3.so.0.0.0
-libxcb-glx.la
libxcb-glx.so
libxcb-glx.so.0
libxcb-glx.so.0.0.0
-libxcb-present.la
libxcb-present.so
libxcb-present.so.0
libxcb-present.so.0.0.0
-libxcb-randr.la
libxcb-randr.so
libxcb-randr.so.0
libxcb-randr.so.0.1.0
-libxcb-record.la
libxcb-record.so
libxcb-record.so.0
libxcb-record.so.0.0.0
-libxcb-render.la
libxcb-render.so
libxcb-render.so.0
libxcb-render.so.0.0.0
-libxcb-res.la
libxcb-res.so
libxcb-res.so.0
libxcb-res.so.0.0.0
-libxcb-screensaver.la
libxcb-screensaver.so
libxcb-screensaver.so.0
libxcb-screensaver.so.0.0.0
-libxcb-shape.la
libxcb-shape.so
libxcb-shape.so.0
libxcb-shape.so.0.0.0
-libxcb-shm.la
libxcb-shm.so
libxcb-shm.so.0
libxcb-shm.so.0.0.0
libxcb-static.a
-libxcb-sync.la
libxcb-sync.so
libxcb-sync.so.1
libxcb-sync.so.1.0.0
-libxcb-xevie.la
libxcb-xevie.so
libxcb-xevie.so.0
libxcb-xevie.so.0.0.0
-libxcb-xf86dri.la
libxcb-xf86dri.so
libxcb-xf86dri.so.0
libxcb-xf86dri.so.0.0.0
-libxcb-xfixes.la
libxcb-xfixes.so
libxcb-xfixes.so.0
libxcb-xfixes.so.0.0.0
-libxcb-xinerama.la
libxcb-xinerama.so
libxcb-xinerama.so.0
libxcb-xinerama.so.0.0.0
-libxcb-xkb.la
libxcb-xkb.so
libxcb-xkb.so.1
libxcb-xkb.so.1.0.0
-libxcb-xprint.la
libxcb-xprint.so
libxcb-xprint.so.0
libxcb-xprint.so.0.0.0
-libxcb-xtest.la
libxcb-xtest.so
libxcb-xtest.so.0
libxcb-xtest.so.0.0.0
-libxcb-xv.la
libxcb-xv.so
libxcb-xv.so.0
libxcb-xv.so.0.0.0
-libxcb-xvmc.la
libxcb-xvmc.so
libxcb-xvmc.so.0
libxcb-xvmc.so.0.0.0
-libxcb.la
libxcb.so
libxcb.so.1
libxcb.so.1.1.0
libz.a
libzmq.a
-libzmq.la
pkgconfig
pkgconfig/Qt5AccessibilitySupport.pc
pkgconfig/Qt5Core.pc
pkgconfig/Qt5DBus.pc
pkgconfig/Qt5DeviceDiscoverySupport.pc
pkgconfig/Qt5EventDispatcherSupport.pc
pkgconfig/Qt5FbSupport.pc
```
</details>

**Review by chefsaroar — APPROVED** [review]

**Comment by dongcarl** [comment]
Updated to also enforce the stricter `--disable-malloc0returnsnull` on `libX11`. Many thanks to @fanquake for his diligent review!

**Review by fanquake — APPROVED** [review]
re-tACK 6ef57ad

**Comment by dongcarl** [comment]
Update: Rebase on #16041.

Sorry for updating this so often, want to make sure that once people with merge access approve, the whole process is smooth.

**Comment by laanwj** [comment]
ACK 8541cbea2, code changes look good to me

**Pull request merged**
